go/cmd/dolt/commands/blame.go

```diff
@@ -0,0 +1,450 @@
+	Author string
+
+	// Description is the description of the commit which last modified the row
+	Description string
+
+	// Timestamp is the timestamp of the commit which last modified the row
+	Timestamp uint64
+}
+
+// TimestampTime returns a time.Time object representing the blameInfo timestamp
+func (bi *blameInfo) TimestampTime() time.Time {
+	return time.Unix(int64(bi.Timestamp/1000), 0)
+}
+
+// TimestampString returns a string representing the blameInfo timestamp
+func (bi *blameInfo) TimestampString() string {
+	return bi.TimestampTime().Format(time.UnixDate)
+}
+
+// A blame graph is a map of primary key hashes to blameInfo structs
+type blameGraph map[hash.Hash]blameInfo
+
+// Blame implements the `dolt blame` command. Blame annotates each row in the given table with information
+// from the revision which last modified the row, optionally starting from a given revision.
+//
+// Blame is computed as follows:
+//
+// First, a blame graph is initialized with one node for every row in the table at the given commit (defaulting
+// to HEAD of the currently checked-out branch).
+//
+// Starting from the given commit, walk backwards through the commit graph (currently by following each commit's
+// first parent, though this may change in the future).
+//
+// For each adjacent pair of commits `old` and `new`, check each remaining unblamed node to see if the row it represents
+// changed between the commits. If so, mark it with `new` as the blame origin and continue to the next node without blame.
+//
+// When all nodes have blame information, stop iterating through commits and print the blame graph.
+func Blame(ctx context.Context, commandStr string, args []string, dEnv *env.DoltEnv) int {
+	ap := argparser.NewArgParser()
+	help, usage := cli.HelpAndUsagePrinters(commandStr, blameShortDesc, blameLongDesc, blameSynopsis, ap)
+	apr := cli.ParseArgs(ap, args, help)
+
+	if apr.NArg() < 2 {
+		usage()
+		return 1
+	}
+
+	cs, err := parseCommitSpec(dEnv, apr)
+	if err != nil {
+		cli.PrintErr(err)
+		return 1
+	}
+
+	tableName := apr.Arg(1)
+
+	if err := runBlame(ctx, dEnv, cs, tableName); err != nil {
+		cli.PrintErr(err)
+		return 1
+	}
+
+	return 0
+}
+
+func runBlame(ctx context.Context, dEnv *env.DoltEnv, cs *doltdb.CommitSpec, tableName string) error {
+	commit, err := dEnv.DoltDB.Resolve(ctx, cs)
+	if err != nil {
+		return err
+	}
+
+	blameGraph, err := blameGraphFromCommit(ctx, dEnv, commit, tableName)
+	if err != nil {
+		return err
+	}
+
+	pkColNames, err := pkColNamesFromCommit(ctx, commit, tableName)
+	if err != nil {
+		return err
+	}
+
+	cli.Println(blameGraph.String(ctx, pkColNames))
+	return nil
+}
+
+func blameGraphFromCommit(ctx context.Context, dEnv *env.DoltEnv, commit *doltdb.Commit, tableName string) (*blameGraph, error) {
+	// get the commits sorted from newest to oldest ending with `commit`
+	commits, err := actions.TimeSortedCommits(ctx, dEnv.DoltDB, commit, -1)
```

Comment:
Do we want time sorted or topological here?

Reply:
Hah, now that I'm looking at it again, is there even a need to get the full list of commits ahead of time? Could probably just keep getting the first parent and using that as the child for the next iteration, right?

Reply:
Maybe, but I think we want to walk non-first-parents too, even if we only compare against first parent in the case of a merge commit?

Reply:
Makes sense.

Reply:
What do you think of the temporal topological sort described on here? https://pvigier.github.io/2019/05/06/commit-graph-drawing-algorithms.html

Reply:
Yeah, I think a topological order with concurrent commits ordered by timestamp is absolutely the thing we want.

This code does a topological sort with ties broken by "height of the tree", and some extra stuff for excluding certain commits from the walk:

https://github.com/liquidata-inc/dolt/blob/master/go/libraries/doltcore/env/actions/commitwalk/commitwalk.go#L133

But it might be the basis for a simpler thing that breaks ties with commit timestamp. (I think basically rip out `SetInvisible` and change `height` and `height` comparisons to a timestamp).

Reply:
Thinking about this further, the reason it's currently height instead of timestamp is that that's the way it gets a topological sort...it needs to be dfs as well, which it isn't in the existing code because it's trying to bail early in the two dot log case.

Reply:
Okay, I think this is right: https://github.com/liquidata-inc/dolt/pull/157/commits/10cec98af42ffadaebdea442a7681c2e80638e84